### PR TITLE
OpenVINO backend

### PR DESCRIPTION
**Pull request by likholat**
**OpenVINO backend for the RIM and Unet2D models.**

Current state:
- `previous_state` and `loglikelihood_scaling` parameters aren't supported for RIM OpenVINO model 
- `FFT` and `IFFT` OpenVINO layers aren't support `normalized` parameter

To do:
- [x] Tune and report OV performance

### CODE REVIEW

**Comment by georgeyiasemis** [comment]
Thanks @likholat for your time and this contribution! Looking forward for more of your work!

**Comment by likholat** [comment]
### Performance results

 Shape| Origin time (ms) | OpenVino time (ms)
:--: | :--: | :--:
[3, 3, 16, 16] | 8.63 | 1.65 **(x5.23)**
[2, 5, 16, 32] | 8.42 | 1.84 **(x4.58)**
[2, 5, 64, 128] | 18.52 | 12.73 **(x1.45)**

Measured on `Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz` with `PyTorch 1.7.0`.


A code similar to the `test_rim.py` was used for the measurement: 
```python
import torch
import time

from direct.data.transforms import fft2, ifft2
from direct.nn.rim.rim import RIM
from direct.nn.openvino.openvino_model import OpenVINOModel
from direct.nn.rim.tests.test_rim import create_input


def main():
    shape = [2, 5, 64, 128]
    n = 30

    model = RIM(
        fft2,
        ifft2,
        hidden_channels=4,
        length=3,
        depth=2,
        no_parameter_sharing=True,
        instance_norm=True,
        dense_connect=True,
        skip_connections=True,
        image_initialization="input-kspace",
    ).cpu()

    img = create_input([shape[0]] + shape[2:] + [2]).cpu()
    kspace = create_input(shape + [2]).cpu()
    sens = create_input(shape + [2]).cpu()
    mask = create_input([shape[0]] + [1] + shape[2:] + [1]).round().int().cpu()

    out = model(img, kspace, mask, sens)

    start = time.time()
    for i in range(n):
        model(img, kspace, mask, sens)
    torch_time = (time.time() - start) / n

    ov_model = OpenVINOModel(model)
    ov_out = ov_model(img, kspace, mask, sens)

    start = time.time()
    for i in range(n):
        ov_model(img, kspace, mask, sens)
    ov_time = (time.time() - start) / n

    print("Torch time:", f"{torch_time * 1000:.{3}f}", "ms")
    print("OV time:", f"{ov_time * 1000:.{3}f}", "ms")

    print('Res diff:')
    print(torch.max(torch.abs(out[0][-1] - ov_out[0][-1])))
    print(torch.max(torch.abs(out[1] - ov_out[1])))


if __name__ == "__main__":
    main()
```

**Comment by likholat** [comment]
@georgeyiasemis, where can I find an example using these models to understand what size of input data is actually relevant?

### Performance results
#### RIM model

 Shape| Origin time (ms) | OpenVino time (ms)
:--: | :--: | :--:
[2, 3, 11, 12] | 9.901 ms | 1.422 ms **(x)**
[2, 3, 65, 128] | 21.823 ms |  13.999 ms **(x)**

#### UNET model

 Shape| Origin time (ms) | OpenVino time (ms)
:--: | :--: | :--:
[4, 5, 16, 32] | 3.430 ms | 0.800 ms **(x)**
[4, 5, 64, 128] | 4.867 ms | 3.229 ms **(x)**

Measured on `Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz` with `PyTorch 1.10.0`.


A code similar to the `test_rim.py` and `test_unet_2d.py` was used for the measurement: 
```python
# example for RIM model

import torch
import time
import argparse

from direct.nn.rim.rim import RIM
from direct.nn.openvino.openvino_model import OpenVINOModel
from tests.tests_nn.test_rim import create_input
from direct.data.transforms import fft2, ifft2


def main():
    parser = argparse.ArgumentParser(description="Generate model and test data")
    parser.add_argument("--shape", type=int, nargs="+", default=[2, 3, 11, 12])
    args = parser.parse_args()

    n = 50

    shape = args.shape
    input_image_is_None = False

    model = RIM(
        fft2,
        ifft2,
        hidden_channels=4,
        length=3,
        depth=2,
        no_parameter_sharing=False,
        instance_norm=True,
        dense_connect=False,
        skip_connections=True,
        image_initialization="zero_filled",
        learned_initializer=True,
    ).cpu()

    inputs = {
        "input_image": create_input([shape[0]] + shape[2:] + [2]).cpu() if not input_image_is_None else None,
        "masked_kspace": create_input(shape + [2]).cpu(),
        "sensitivity_map": create_input(shape + [2]).cpu(),
        "sampling_mask": create_input([shape[0]] + [1] + shape[2:] + [1]).round().int().cpu(),
    }

    out = model(**inputs)

    start = time.time()
    for i in range(n):
        model(**inputs)
    torch_time = (time.time() - start) / n

    ov_model = OpenVINOModel(model)
    ov_out = ov_model(**inputs)

    start = time.time()
    for i in range(n):
        ov_model(**inputs)
    ov_time = (time.time() - start) / n

    print("Torch time:", f"{torch_time * 1000:.{3}f}", "ms")
    print("OV time:", f"{ov_time * 1000:.{3}f}", "ms")

    print("Res diff:")
    print(torch.max(torch.abs(out[0][-1] - ov_out[0][-1])))
    print(torch.max(torch.abs(out[1] - ov_out[1])))


if __name__ == "__main__":
    main()
```

**Comment by likholat** [comment]
@georgeyiasemis, please review.

Is it critical that I changed `numpy>=1.21.2` to `numpy<1.20`?

**Comment by georgeyiasemis** [comment]
> @georgeyiasemis, please review.

Hi @likholat, thanks for your contribution. We are currently reviewing #172, hence this might need to be rebased to your branch.  

> Is it critical that I changed `numpy>=1.21.2` to `numpy<1.20`?

We prefer to use the most updated versions of the model. Could you please point me to _how_ this will affect your work?

**Comment by likholat** [comment]
> > @georgeyiasemis, please review.
> 
> Hi @likholat, thanks for your contribution. We are currently reviewing #172, hence this might need to be rebased to your branch.
> 
> > Is it critical that I changed `numpy>=1.21.2` to `numpy<1.20`?
> 
> We prefer to use the most updated versions of the modules. Could you please point me to _how_ this will affect your work?

`openvino 2021.4.2` depends on `numpy<1.20`

**Comment by georgeyiasemis** [comment]
Hi @likholat. You can now rebase from main.

Can you also please give me an example of how we can train an openvivo backened model in DIRECT?

**Comment by likholat** [comment]
@georgeyiasemis, OpenVINO is only used for pretrained models.

**Comment by likholat** [comment]
@georgeyiasemis, please review again.

File: direct/data/transforms.py [hunk 1]
```diff
@@ -16,6 +16,39 @@
...
+
+    @staticmethod
+    def forward(self, data, dim, centered, normalized, inverse=False):
+        if inverse:
+            return origin_ifft2(data, dim, centered, normalized)
+        else:
+            return origin_fft2(data, dim, centered, normalized)
+
+
+class ComplexMultiplicationONNX(torch.autograd.Function):
```
Comment (georgeyiasemis) [inline direct/data/transforms.py:37]: Please derive all ONNX classes from a single class, e.g:
`class ONNX(torch.autograd.Function):
   ....
class ComplexMultiplicationONNX(ONNX):
   ....
`

File: direct/data/transforms.py [hunk 2]
```diff
@@ -153,7 +186,20 @@ def fft2(
...
+    data: torch.Tensor,
+    dim: Tuple[int, ...] = (1, 2),
+    centered: bool = True,
+    normalized: bool = True,
+) -> torch.Tensor:
+
+    if data.requires_grad:
+        return origin_fft2(data, dim, centered, normalized)
+    else:
+        return FFTONNX.apply(data, dim, centered, normalized)
```
Comment (georgeyiasemis) [inline direct/data/transforms.py:199]: Please provide an example of a reconstructed image with our `fft2`/`ifft2` torch function compared to a reconstructed image with `FFTONNX`.  

What does `torch.allclose(origin_fft2(data, dim, centered, normalized), FFTONNX.apply(data, dim, centered, normalized))` returns?

File: direct/nn/openvino/openvino_model.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
+from openvino_extensions import get_extensions_path
+from openvino.inference_engine import IECore
+
+import torch
+import torch.nn as nn
+import io
```
Comment (georgeyiasemis) [inline direct/nn/openvino/openvino_model.py:6]: Please use `isort`.

File: direct/nn/openvino/openvino_model.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        return out
+
+    def forward(self, *args, **kwargs):
+        if self.exec_net is None:
+            self.create_net(*args, **kwargs)
+
+        ov_input = dict(zip(self.input_names, self.inputs))
+        res = self.exec_net.infer(ov_input)
+
+        return self.postprocess(res)
```
Comment (georgeyiasemis) [inline direct/nn/openvino/openvino_model.py:107]: Please make sure that all new methods/classes have docstrings and are adequately documented.

File: direct/nn/recurrent/recurrent.py [hunk 5]
```diff
@@ -145,14 +145,13 @@ def forward(
...
+            stacked_inputs = torch.cat([cell_input, prev_state], dim=1)
 
             update = torch.sigmoid(self.update_gates[idx](stacked_inputs))
             reset = torch.sigmoid(self.reset_gates[idx](stacked_inputs))
-            delta = torch.tanh(
-                self.out_gates[idx](torch.cat([cell_input, previous_state[:, :, :, :, idx] * reset], dim=1))
-            )
-            cell_input = previous_state[:, :, :, :, idx] * (1 - update) + delta * update
+            delta = torch.tanh(self.out_gates[idx](torch.cat([cell_input, prev_state * reset], dim=1)))
+            cell_input = prev_state * (1 - update) + delta * update
```
Comment (georgeyiasemis) [inline direct/nn/recurrent/recurrent.py:154]: Why was that necessary?

File: direct/nn/unet/unet_2d.py [hunk 6]
```diff
@@ -299,7 +299,7 @@ def unpad(
         w_mult: int,
     ) -> torch.Tensor:
 
-        return input_data[..., h_pad[0] : h_mult - h_pad[1], w_pad[0] : w_mult - w_pad[1]]
+        return input_data[..., int(h_pad[0]) : int(h_mult - h_pad[1]), int(w_pad[0]) : int(w_mult - w_pad[1])]
```
Comment (georgeyiasemis) [inline direct/nn/unet/unet_2d.py:302]: Why was that necessary?

**Review by georgeyiasemis — COMMENTED** [review]
@likholat please take a look at my comments.

File: direct/data/transforms.py (hunk 2 above)
Comment (likholat) [reply direct/data/transforms.py:199]: Which demo can be used to provide an example of a reconstructed image?

Can I use a random `data` for comparison?

File: direct/nn/recurrent/recurrent.py (hunk 5 above)
Comment (likholat) [reply direct/nn/recurrent/recurrent.py:154]: This reduces the running time of the OpenVINO model.

File: direct/nn/unet/unet_2d.py (hunk 6 above)
Comment (likholat) [reply direct/nn/unet/unet_2d.py:302]: This is necessary to convert to OpenVINO model from onnx.

**Comment by likholat** [comment]
@georgeyiasemis, please review again. Replied and fixed all comments.

File: direct/data/transforms.py (hunk 2 above)
Comment (likholat) [reply direct/data/transforms.py:210]:
```python
import torch
from direct.data import transforms as T
from tests.tests_nn.test_rim import create_input

def main():
    inp = create_input([1, 3, 64, 64, 2])

    ref = T.origin_fft2(inp, (1, 2), True, True)
    out = T.FFTONNX.apply(inp, (1, 2), True, True)

    print(torch.allclose(ref, out)) # True

if __name__ == "__main__":
    main()
```

I use the script above for comparison and `torch.allclose(ref, out)` returns `True`.

**Comment by georgeyiasemis** [comment]
@likholat Could you please check the code coverage report (it dropped -0.72%) and implement test for the code you added? Then we are good to merge.

**Review by georgeyiasemis — COMMENTED** [review]
Ok

**Comment by likholat** [comment]
Thanks for review! I will help with maintenance of OpenVINO related code so feel free to ping in case of issues.

File: direct/data/transforms.py (hunk 1 above)
Comment (likholat) [reply direct/data/transforms.py]: Done

File: direct/data/transforms.py (hunk 2 above)
Comment (likholat) [reply direct/data/transforms.py]: Done

File: direct/nn/openvino/openvino_model.py (hunk 3 above)
Comment (likholat) [reply direct/nn/openvino/openvino_model.py]: Done

File: direct/nn/openvino/openvino_model.py (hunk 4 above)
Comment (likholat) [reply direct/nn/openvino/openvino_model.py]: Done

**Comment by dkurt** [comment]
Hi, @georgeyiasemis! Not sure why, but I don't see OpenVINO anymore in the project. For example, there is no added test from this PR here: https://github.com/NKI-AI/direct/blob/main/tests/tests_nn/test_unet_2d.py. Maybe main branch was force pushed?
